File: IOS_CampusKinect/CampusKinect_IOS/Core/Services/TermsOfServiceManager.swift
import Foundation

class TermsOfServiceManager: ObservableObject {
    static let shared = TermsOfServiceManager()
    
    @Published var shouldShowTerms = false
    
    private let userDefaults = UserDefaults.standard
    private let hasAcceptedTermsKey = "hasAcceptedTerms"
    private let shouldRememberChoiceKey = "shouldRememberTermsChoice"
    private let termsVersionKey = "acceptedTermsVersion"
    
    // Current terms version - increment when terms are updated
    private let currentTermsVersion = "1.0"
    
    private init() {}
    
    // MARK: - Public Methods
    
    /// Check if user needs to see terms popup
    func shouldShowTermsPopup(for userId: String) -> Bool {
        let userSpecificKey = "\(hasAcceptedTermsKey)_\(userId)"
        let userSpecificRememberKey = "\(shouldRememberChoiceKey)_\(userId)"
        let userSpecificVersionKey = "\(termsVersionKey)_\(userId)"
        
        // Check if user has accepted terms
        let hasAccepted = userDefaults.bool(forKey: userSpecificKey)
        let shouldRemember = userDefaults.bool(forKey: userSpecificRememberKey)
        let acceptedVersion = userDefaults.string(forKey: userSpecificVersionKey)
        
        // Show terms if:
        // 1. User has never accepted terms, OR
        // 2. User chose not to remember their choice, OR
        // 3. Terms version has been updated
        return !hasAccepted || !shouldRemember || acceptedVersion != currentTermsVersion
    }
    
    /// Record that user has accepted terms
    func acceptTerms(for userId: String, shouldRememberChoice: Bool) {
        let userSpecificKey = "\(hasAcceptedTermsKey)_\(userId)"
        let userSpecificRememberKey = "\(shouldRememberChoiceKey)_\(userId)"
        let userSpecificVersionKey = "\(termsVersionKey)_\(userId)"
        
        userDefaults.set(true, forKey: userSpecificKey)
        userDefaults.set(shouldRememberChoice, forKey: userSpecificRememberKey)
        userDefaults.set(currentTermsVersion, forKey: userSpecificVersionKey)
        
        print("📋 Terms accepted for user \(userId), remember choice: \(shouldRememberChoice)")
    }
    
    /// Check if terms should be shown for current session
    func checkAndShowTermsIfNeeded(for userId: String) {
        shouldShowTerms = shouldShowTermsPopup(for: userId)
        if shouldShowTerms {
            print("📋 Terms popup will be shown for user \(userId)")
        } else {
            print("📋 Terms popup not needed for user \(userId)")
        }
    }
    
    /// Force show terms (for testing or terms updates)
    func forceShowTerms() {
        shouldShowTerms = true
    }
    
    /// Reset terms acceptance for a user (for testing)
    func resetTermsAcceptance(for userId: String) {
        let userSpecificKey = "\(hasAcceptedTermsKey)_\(userId)"
        let userSpecificRememberKey = "\(shouldRememberChoiceKey)_\(userId)"
        let userSpecificVersionKey = "\(termsVersionKey)_\(userId)"
        
        userDefaults.removeObject(forKey: userSpecificKey)
        userDefaults.removeObject(forKey: userSpecificRememberKey)
        userDefaults.removeObject(forKey: userSpecificVersionKey)
        
        print("📋 Terms acceptance reset for user \(userId)")
    }
    
    /// Reset terms acceptance for all users (for Apple review preparation)
    func resetAllTermsAcceptance() {
        // Get all UserDefaults keys and remove terms-related ones
        let allKeys = userDefaults.dictionaryRepresentation().keys
        let termsKeys = allKeys.filter { key in
            key.contains(hasAcceptedTermsKey) || 
            key.contains(shouldRememberChoiceKey) || 
            key.contains(termsVersionKey)
        }
        
        for key in termsKeys {
            userDefaults.removeObject(forKey: key)
        }
        
        print("📋 All terms acceptance data cleared - Terms popup will show for all users")
    }
    
    /// Get terms acceptance status for debugging
    func getTermsStatus(for userId: String) -> (hasAccepted: Bool, shouldRemember: Bool, version: String?) {
        let userSpecificKey = "\(hasAcceptedTermsKey)_\(userId)"
        let userSpecificRememberKey = "\(shouldRememberChoiceKey)_\(userId)"
        let userSpecificVersionKey = "\(termsVersionKey)_\(userId)"
        
        return (
            hasAccepted: userDefaults.bool(forKey: userSpecificKey),
            shouldRemember: userDefaults.bool(forKey: userSpecificRememberKey),
            version: userDefaults.string(forKey: userSpecificVersionKey)
        )
    }
} 

File: Web_CampusKinect/src/services/termsOfServiceManager.ts
class TermsOfServiceManager {
  private static instance: TermsOfServiceManager;
  
  private readonly hasAcceptedTermsKey = 'hasAcceptedTerms';
  private readonly shouldRememberChoiceKey = 'shouldRememberTermsChoice';
  private readonly termsVersionKey = 'acceptedTermsVersion';
  
  // Current terms version - increment when terms are updated
  private readonly currentTermsVersion = '1.0';
  
  private constructor() {}
  
  static getInstance(): TermsOfServiceManager {
    if (!TermsOfServiceManager.instance) {
      TermsOfServiceManager.instance = new TermsOfServiceManager();
    }
    return TermsOfServiceManager.instance;
  }
  
  /**
   * Check if user needs to see terms popup
   */
  shouldShowTermsPopup(userId: string): boolean {
    const userSpecificKey = `${this.hasAcceptedTermsKey}_${userId}`;
    const userSpecificRememberKey = `${this.shouldRememberChoiceKey}_${userId}`;
    const userSpecificVersionKey = `${this.termsVersionKey}_${userId}`;
    const sessionShownKey = `termsShownThisSession_${userId}`;
    
    // Check if user has accepted terms
    const hasAccepted = localStorage.getItem(userSpecificKey) === 'true';
    const shouldRemember = localStorage.getItem(userSpecificRememberKey) === 'true';
    const acceptedVersion = localStorage.getItem(userSpecificVersionKey);
    const shownThisSession = sessionStorage.getItem(sessionShownKey) === 'true';
    
    // Show terms if:
    // 1. User has never accepted terms, OR
    // 2. Terms version has been updated, OR
    // 3. User chose "Show Every Login" AND terms haven't been shown this session
    if (!hasAccepted || acceptedVersion !== this.currentTermsVersion) {
      return true;
    }
    
    // If user chose "Remember Choice", never show again
    if (shouldRemember) {
      return false;
    }
    
    // If user chose "Show Every Login", only show if not shown this session
    return !shownThisSession;
  }
  
  /**
   * Record that user has accepted terms
   */
  acceptTerms(userId: string, shouldRememberChoice: boolean): void {
    const userSpecificKey = `${this.hasAcceptedTermsKey}_${userId}`;
    const userSpecificRememberKey = `${this.shouldRememberChoiceKey}_${userId}`;
    const userSpecificVersionKey = `${this.termsVersionKey}_${userId}`;
    const sessionShownKey = `termsShownThisSession_${userId}`;
    
    localStorage.setItem(userSpecificKey, 'true');
    localStorage.setItem(userSpecificRememberKey, shouldRememberChoice.toString());
    localStorage.setItem(userSpecificVersionKey, this.currentTermsVersion);
    
    // Mark that terms have been shown in this session (for "Show Every Login" users)
    sessionStorage.setItem(sessionShownKey, 'true');
    
    console.log(`📋 Terms accepted for user ${userId}, remember choice: ${shouldRememberChoice}`);
  }
  
  /**
   * Reset terms acceptance for a user (for testing)
   */
  resetTermsAcceptance(userId: string): void {
    const userSpecificKey = `${this.hasAcceptedTermsKey}_${userId}`;
    const userSpecificRememberKey = `${this.shouldRememberChoiceKey}_${userId}`;
    const userSpecificVersionKey = `${this.termsVersionKey}_${userId}`;
    const sessionShownKey = `termsShownThisSession_${userId}`;
    
    localStorage.removeItem(userSpecificKey);
    localStorage.removeItem(userSpecificRememberKey);
    localStorage.removeItem(userSpecificVersionKey);
    sessionStorage.removeItem(sessionShownKey);
    
    console.log(`📋 Terms acceptance reset for user ${userId}`);
  }
  
  /**
   * Reset terms acceptance for all users (for testing)
   */
  resetAllTermsAcceptance(): void {
    const localKeys = Object.keys(localStorage);
    const termsKeys = localKeys.filter(key => 
      key.includes(this.hasAcceptedTermsKey) || 
      key.includes(this.shouldRememberChoiceKey) || 
      key.includes(this.termsVersionKey)
    );
    
    const sessionKeys = Object.keys(sessionStorage);
    const sessionTermsKeys = sessionKeys.filter(key => 
      key.includes('termsShownThisSession')
    );
    
    termsKeys.forEach(key => localStorage.removeItem(key));
    sessionTermsKeys.forEach(key => sessionStorage.removeItem(key));
    
    console.log('📋 All terms acceptance data cleared - Terms popup will show for all users');
  }
  
  /**
   * Get terms acceptance status for debugging
   */
  getTermsStatus(userId: string): { hasAccepted: boolean; shouldRemember: boolean; version: string | null; shownThisSession: boolean } {
    const userSpecificKey = `${this.hasAcceptedTermsKey}_${userId}`;
    const userSpecificRememberKey = `${this.shouldRememberChoiceKey}_${userId}`;
    const userSpecificVersionKey = `${this.termsVersionKey}_${userId}`;
    const sessionShownKey = `termsShownThisSession_${userId}`;
    
    return {
      hasAccepted: localStorage.getItem(userSpecificKey) === 'true',
      shouldRemember: localStorage.getItem(userSpecificRememberKey) === 'true',
      version: localStorage.getItem(userSpecificVersionKey),
      shownThisSession: sessionStorage.getItem(sessionShownKey) === 'true'
    };
  }
}

export const termsOfServiceManager = TermsOfServiceManager.getInstance(); 